tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/tracee/UJsonTraceReader.scala

```diff
@@ -0,0 +1,94 @@
+package at.forsyte.apalache.tla.tracee
+
+import at.forsyte.apalache.infra.passes.options.SourceOption
+import at.forsyte.apalache.infra.passes.options.SourceOption.FileSource
+import at.forsyte.apalache.io.json.impl.{UJsonRep, UJsonScalaFactory, UJsonToTlaViaBuilder}
+import at.forsyte.apalache.infra.passes.options.SourceOption._
+import at.forsyte.apalache.io.json.JsonDeserializationError
+import at.forsyte.apalache.io.lir.TypeTagReader
+import at.forsyte.apalache.tla.imp.src.SourceStore
+import at.forsyte.apalache.tla.lir.oper.{TlaBoolOper, TlaOper}
+import at.forsyte.apalache.tla.lir.{NameEx, OperEx, TlaEx, TlaOperDecl}
+
+import scala.annotation.unused
+import scala.util.{Failure, Success, Try}
+
+/**
+ * Reads an ITF-format or a TlaJSON-format trace into a TLAIR trace, with an intermediate UJson representation.
+ *
+ * @author
+ *   Jure Kukovec
+ */
+class UJsonTraceReader(sourceStoreOpt: Option[SourceStore], tagReader: TypeTagReader) extends TraceReader[UJsonRep] {
+  private val builder = new UJsonToTlaViaBuilder(sourceStoreOpt)(tagReader)
+  override def read(source: SourceOption): UJsonRep = {
+    val readable: ujson.Readable = source match {
+      case FileSource(f, Format.Json) => f
+      case _                          => throw new IllegalArgumentException("File provided with --trace is not a JSON.")
+    }
+
+    // Rethrow as JsonDeserializationError if unable to read
+    Try(ujson.read(readable)) match {
+      case Success(ujsonVal) => UJsonRep(ujsonVal)
+      case Failure(exception) =>
+        throw new JsonDeserializationError("Unable to read --trace as JSON.", exception)
+    }
+  }
+
+  // TODO
+  private def convertITF(@unused json: UJsonRep): Trace = IndexedSeq.empty
+
+  private def kvFromAsgn(ex: TlaEx): (String, TlaEx) = ex match {
+    case OperEx(TlaOper.eq, NameEx(name), rhs) => name -> rhs
+    case _ => throw new JsonDeserializationError(s"Cannot read variable assignment from $ex.")
+  }
+
+  private def convertGenericJSON(json: UJsonRep): Trace = {
+    val operDecls = for {
+      modules <- json.getFieldOpt("modules")
+      decls <- UJsonScalaFactory.asSeq(modules).head.getFieldOpt("declarations")
+    } yield {
+      // drop CInit (head) and Inv (last)
+      UJsonScalaFactory.asSeq(decls).tail.dropRight(1).toIndexedSeq.map { decl =>
+        builder.asTlaDecl(decl).asInstanceOf[TlaOperDecl]
+      }
+    }
+    // From an indexed sequence of declarations, create a sequence of State-maps.
+    operDecls
+      .map { decls =>
+        decls.map { decl =>
+          decl.body match {
+            // Edge case: single-var specs
+            case OperEx(TlaOper.eq, NameEx(name), rhs) => Map(name -> rhs)
+            // general conjunctive case
```

Comment:
This looks like `kvFromAsgn`. A bit strange to me that we have the same logic here and in that abstraction... perhaps you should replace `kvFromAsgn` with a function performing this transformation into maps (lines 62-68)?

Reply:
The reason I don't call `kvFromAsgn` here again is that you already do the pattern matching that happens inside `kvFromAsgn` beforehand, whereas in the following lines, you map `kvFromAsgn` over the contents of a sequence of `TlaEx`, the shapes of which are unknown.